### Loading dataframes

In [1]:
import pandas as pd

### Check for duplicates

In [80]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,20,30,40,50]})

In [81]:
df

a   b
0  1  10
1  1  10
2  1  11
3  2  20
4  2  20
5  3  30
6  4  40
7  5  50

In [82]:
df.duplicated()

0    False
1     True
2    False
3    False
4     True
5    False
6    False
7    False
dtype: bool

In [83]:
df.duplicated(subset=['a'])

0    False
1     True
2     True
3    False
4     True
5    False
6    False
7    False
dtype: bool

In [84]:
df.duplicated().sum()

2

### Drop duplicated rows

In [86]:
df_no_duplicates = df.drop_duplicates()

In [87]:
len(df), len(df_no_duplicates)

(8, 6)

In [88]:
df.drop_duplicates(inplace=True)

In [89]:
len(df)

6

### Check for missing values

In [131]:
df = pd.DataFrame({'a':[1,2,3,4],
                  'b':[10,None,30,40],
                  'c':[None,None,None,400]})

In [114]:
df.isnull()

a      b      c
0  False  False   True
1  False   True   True
2  False  False   True
3  False  False  False

In [115]:
df.isnull().any(axis=0)

a    False
b     True
c     True
dtype: bool

In [116]:
df.isnull().any(axis=1)

0     True
1     True
2     True
3    False
dtype: bool

In [132]:
df.isnull().sum()

a    0
b    1
c    3
dtype: int64

### Drop rows with missing values

In [117]:
df

a     b      c
0  1  10.0    NaN
1  2   NaN    NaN
2  3  30.0    NaN
3  4  40.0  400.0

In [119]:
df_only_full_rows = df.dropna()
df_only_full_rows

a     b      c
3  4  40.0  400.0

In [120]:
df_rows_where_b_is_not_missing = df.dropna(subset=['b'])
df_rows_where_b_is_not_missing

a     b      c
0  1  10.0    NaN
2  3  30.0    NaN
3  4  40.0  400.0

### Replace missing values

In [121]:
df

a     b      c
0  1  10.0    NaN
1  2   NaN    NaN
2  3  30.0    NaN
3  4  40.0  400.0

In [124]:
mean_b = df['b'].mean()
df_missing_b_replaced_with_mean = df.fillna(value={'b':mean_b})
df_missing_b_replaced_with_mean

a          b      c
0  1  10.000000    NaN
1  2  26.666667    NaN
2  3  30.000000    NaN
3  4  40.000000  400.0

In [125]:
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero = df.fillna(value={'b':mean_b,
                                                                                'c':0})
df_missing_b_replaced_with_mean_missing_c_replaced_with_zero

a          b      c
0  1  10.000000    0.0
1  2  26.666667    0.0
2  3  30.000000    0.0
3  4  40.000000  400.0

In [126]:
df.fillna('Unknown', inplace=True)
df

a        b        c
0  1       10  Unknown
1  2  Unknown  Unknown
2  3       30  Unknown
3  4       40      400

### Describing numeric data

In [162]:
df = pd.DataFrame({'a':[1,1,1,2,2,3,4,5],
                  'b':[10,10,11,20,None,None,40,50],
                  'c':['apple','apple','plum','pear','plum','apple','apple','apple']})

In [163]:
df

a     b      c
0  1  10.0  apple
1  1  10.0  apple
2  1  11.0   plum
3  2  20.0   pear
4  2   NaN   plum
5  3   NaN  apple
6  4  40.0  apple
7  5  50.0  apple

In [164]:
df.describe()

a          b
count  8.000000   6.000000
mean   2.375000  23.500000
std    1.505941  17.363755
min    1.000000  10.000000
25%    1.000000  10.250000
50%    2.000000  15.500000
75%    3.250000  35.000000
max    5.000000  50.000000

In [165]:
df.max()

a       5
b      50
c    plum
dtype: object

In [166]:
df['a'].max()

5

### Describing non numeric data

In [167]:
df

a     b      c
0  1  10.0  apple
1  1  10.0  apple
2  1  11.0   plum
3  2  20.0   pear
4  2   NaN   plum
5  3   NaN  apple
6  4  40.0  apple
7  5  50.0  apple

In [168]:
df['c'].unique()

array(['apple', 'plum', 'pear'], dtype=object)

In [169]:
df['c'].nunique()

3

In [172]:
df['c'].value_counts()

apple    5
plum     2
pear     1
Name: c, dtype: int64

## Exercise

### 1 - exercise
Load the datasets into pandas dataframes called trip, weather and station. (Don't forget to import the pandas library first!) <br>
Create three variables called trip_duplicates_num, weather_duplicates_num and station_duplicatess_num that contains how many duplicated rows are in each dataframe.

In [2]:
### Your code here
import pandas as pd

weather = pd.read_csv('weather.csv')
trip = pd.read_csv('trip.csv')
station = pd.read_csv('station.csv')

trip_duplicates_num = trip.duplicated().sum()
station_duplicates_num = station.duplicated().sum()
weather_duplicates_num = weather.duplicated().sum()

/home/esztersomos/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2698: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### 1 - check yourself

In [191]:
print('Length of trip dataframe should be 144115 with 100 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(trip), trip_duplicates_num))
print('Length of weather dataframe should be 928 with 8 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(weather), weather_duplicates_num))
print('Length of station dataframe should be 69 with 0 duplicated rows.\n\
The length of your dataframe is {} and you counted {} duplicated rows\n'.format(len(station), station_duplicates_num))

Length of trip dataframe should be 144115 with 100 duplicated rows.
The length of your dataframe is 144115 and you counted 100 duplicated rows

Length of weather dataframe should be 928 with 8 duplicated rows.
The length of your dataframe is 928 and you counted 8 duplicated rows

Length of station dataframe should be 69 with 0 duplicated rows.
The length of your dataframe is 69 and you counted 0 duplicated rows



### 2 - exercise
For all 3 dataframes delete the duplicated rows in place (without creating a new dataframe) <br>

In [192]:
### Your code here

weather.drop_duplicates(inplace=True)
trip.drop_duplicates(inplace=True)
station.drop_duplicates(inplace=True)

### 2 - check yourself

In [193]:
print('Length of trip dataframe should be 144015.\n\
The length of your dataframe is {}\n'.format(len(trip)))
print('Length of weather dataframe should be 920.\n\
The length of your dataframe is {}\n'.format(len(weather)))
print('Length of station dataframe should be 69.\n\
The length of your dataframe is {}\n'.format(len(station)))

Length of trip dataframe should be 144015.
The length of your dataframe is 144015

Length of weather dataframe should be 920.
The length of your dataframe is 920

Length of station dataframe should be 69.
The length of your dataframe is 69



### 3 - exercise
For all dataframes check if there are columns with missing values. <br>
Create 3 lists called trip_columns_with_missing_data, weather_columns_with_missing_data, station_columns_with_missing_data that contains the names of the columns with missing values of the dataframe. <br>
You can populate these lists by hand, or as an advanced task you can use pandas methods.

In [194]:
### Your code here
weather_columns_with_missing_data = weather.isnull().any()[weather.isnull().any()==True].index
trip_columns_with_missing_data = trip.isnull().any()[trip.isnull().any()==True].index
station_columns_with_missing_data = station.isnull().any()[station.isnull().any()==True].index

### 3 - check yourself

In [195]:
print('Columns with missing values in the trip dataframe are:\n \
Subscription Type,Zip Code\n\
You have found:\n {}\n'.format(','.join(trip_columns_with_missing_data)))

print('Columns with missing values in the weather dataframe are:\n \
Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events\n\
You have found:\n {}\n'.format(','.join(weather_columns_with_missing_data)))

print('Columns with missing values in the station dataframe are:\n \
\n\
You have found:\n {}\n'.format(','.join(station_columns_with_missing_data)))

Columns with missing values in the trip dataframe are:
 Subscription Type,Zip Code
You have found:
 Subscription Type,Zip Code

Columns with missing values in the weather dataframe are:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events
You have found:
 Max_Temperature_F,Mean_Temperature_F,Min_TemperatureF,Max_Gust_Speed_MPH,Events

Columns with missing values in the station dataframe are:
 
You have found:
 



### 4 - exercise
How many values are missing in each column with missing values in the dataframes? <br>
Display the answer in any format you would like. As an advance task, try to display only the names of the columns that have missing values in them.

In [196]:
### Your code here
weather.isnull().sum()[weather.isnull().any()==True]

Max_Temperature_F       3
Mean_Temperature_F      3
Min_TemperatureF        3
Max_Gust_Speed_MPH    138
Events                782
dtype: int64

In [197]:
trip.isnull().sum()[trip.isnull().any()==True]

Subscription Type      10
Zip Code             6130
dtype: int64

### 4 - check yourself
The number of missing values in Max_Temperature_F column is 3 <br>
The number of missing values in Mean_Temperature_F column is 3<br>
The number of missing values in Min_TemperatureF column is 3<br>
The number of missing values in Max_Gust_Speed_MPH column is 138<br>
The number of missing values in Events column is 782<br>
The number of missing values in Subscription Type column is 10<br>
The number of missing values in Zip Code column is 6130<br>

### 5 - exercise
Before deciding how to deal with the missing values, let's get more familiar with the data! <br>
 - How many values are in the columns with missing data? As an advanced task, try to display the number of rows for only these columns, not the others.
 - Display the mean of each numeric column in the weather dataframe. Which are 2 columns with the lowest mean? As an advanced task, try to display them in descending order!
 - And what about the Events column in the weather dataframe? What are the unique values and how many times do they occur?

In [7]:
weather[['Max_Temperature_F','Mean_Temperature_F','Min_TemperatureF','Max_Gust_Speed_MPH','Events']].count()

Max_Temperature_F     925
Mean_Temperature_F    925
Min_TemperatureF      925
Max_Gust_Speed_MPH    790
Events                139
dtype: int64

In [5]:
trip[['Subscription Type','Zip Code']].count()

Subscription Type    144105
Zip Code             137973
dtype: int64

In [199]:
### Your code here
weather.mean().sort_values(ascending=False)

zip                            94325.000000
Wind_Dir_Degrees                 269.507609
Min_TemperatureF                  97.137405
Max_Humidity                      83.628261
Max_Temperature_F                 67.002181
Mean_Humidity                     65.433696
Mean_Temperature_F                56.926936
Max_Dew_Point_F                   48.943478
Min_Humidity                      43.616304
MeanDew_Point_F                   43.552174
Min_Dewpoint_F                    38.241304
Max_Sea_Level_Pressure_In         30.122978
Mean_Sea_Level_Pressure_In        30.065750
Min_Sea_Level_Pressure_In         29.997674
Max_Gust_Speed_MPH                21.919437
Max_Wind_Speed_MPH                15.075000
Max_Visibility_Miles               9.993478
Mean_Visibility_Miles              9.521739
Min_Visibility_Miles               7.851087
Mean_Wind_Speed_MPH                4.290217
Cloud_Cover                        2.143478
dtype: float64

In [209]:
weather['Events'].value_counts()

Rain        101
Fog          34
rain          2
Fog-Rain      1
Name: Events, dtype: int64

### 5 - check yourself
The number of rows without missing data are:
- Max_Temperature_F column is 925
- Mean_Temperature_F column is 925
- Min_TemperatureF column is 925
- Max_Gust_Speed_MPH column is 790
- Events column is 139
- Subscription Type column is 144105
- Zip Code column is 137973 <br>

The columns with the lowest mean values are Cloud_clover and Mean_Wind_Speed_MPH <br><br>
In the Events column there are 101 rows with Rain, 34 rows with Fog, 2 rows with rain and 1 row with Fog-Rain

### 6 - exercise
So let's decide what we will do with the missing data!<br>
- In the Temperature and the Gust Speed columns, there are not too many missing data, so let's fill those cells with the mean of the column.
- In the Events columns, the missing value means that there were no rain or fog that day. So let's fill those cells with the string 'no_event'
- In the Subscription type column, only a few rows are missing and we can't guess the original values. Let's delete those rows from the dataframe.
- In the Zip Code column there are a lot of missing values, but we might be able to guess the original ones based on the street names later. Leave them as missing for now. <br>

Create a new weather and trip dataframe called weather_filled and trip_filled where these solutions are applied!

In [13]:
### Your code here
weather_filled = weather.fillna(value={'Max_Temperature_F': weather.Max_Temperature_F.mean(),
                                        'Mean_Temperature_F': weather.Mean_Temperature_F.mean(),
                                        'Min_TemperatureF': weather.Min_TemperatureF.mean(),
                                        'Max_Gust_Speed_MPH': weather.Max_Gust_Speed_MPH.mean(),
                                        'Events': 'no_event'
                                       })
trip_filled = trip.dropna(subset=['Subscription Type'])

### 6 - check yourself

In [54]:
print('In the weather_filled dataframe the number of missing data should be 0\n \
and in your dataframe the number is {}\n'.format(weather_filled.isnull().any().sum()))
print('The number of rows where the replacing value is not correct in the weather_filled dataframe: \n \
in column Max_Temperature_F is {} \n \
in column Mean_Temperature_F is {} \n \
in column Min_TemperatureF is {} \n \
in column Max_Gust_Speed_MPH is {} \n \
in column Events is {} \n '.format((weather_filled[weather.Max_Temperature_F.isnull()]['Max_Temperature_F'] != weather.Max_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Mean_Temperature_F.isnull()]['Mean_Temperature_F'] != weather.Mean_Temperature_F.mean()).sum(),
                                             (weather_filled[weather.Min_TemperatureF.isnull()]['Min_TemperatureF'] != weather.Min_TemperatureF.mean()).sum(),
                                             (weather_filled[weather.Max_Gust_Speed_MPH.isnull()]['Max_Gust_Speed_MPH'] != weather.Max_Gust_Speed_MPH.mean()).sum(),
                                             (weather_filled[weather.Events.isnull()]['Events'] != 'no_event').sum()))
print('The length of the trip_filled dataframe should be 144105\n \
and in your dataframe number is {}\n'.format(len(trip_filled)))

print('There number of missing values in the Zip Code columns of the trip_filled dataframe should be 6142\n \
and in your dataframe number is {}\n'.format(trip_filled['Zip Code'].isnull().sum()))

In the weather_filled dataframe the number of missing data should be 0
 and in your dataframe the number is 0

The number of rows where the replacing value is not correct in the weather_filled dataframe: 
 in column Max_Temperature_F is 0 
 in column Mean_Temperature_F is 0 
 in column Min_TemperatureF is 0 
 in column Max_Gust_Speed_MPH is 0 
 in column Events is 0 
 
The length of the trip_filled dataframe should be 144105
 and in your dataframe number is 144105

There number of missing values in the Zip Code columns of the trip_filled dataframe should be 6142
 and in your dataframe number is 6142



### 7 - exercise
Save these new dataframes into csv-s called weather_filled.csv and trip_filled.csv!

In [59]:
### Your code here
import os
os.chdir('/home/esztersomos/Documents/CEU/winter_2019/checking_data2/')
weather_filled.to_csv('weather_filled.csv', index=False)
trip_filled.to_csv('trip_filled.csv', index=False)
station.to_csv('station.csv', index=False)

### 7 - check yourself

In [60]:
if 'trip_filled.csv' in os.listdir():
    print('trip_filled.csv was successfully saved')
else:
    print('trip.csv was NOT successfully saved')
if 'weather_filled.csv' in os.listdir():
    print('weather_filled.csv was successfully saved')
else:
    print('weather_filled.csv was NOT successfully saved')

trip_filled.csv was successfully saved
weather_filled.csv was successfully saved
